### PR TITLE
Add image segmentation docs

### PR DESCRIPTION
**Pull request by mattcai**
Added tutorials for watershed, manual, and ilastik segmentation in starfish. Also made minor edits to `from_fiji_roi_sets` docstring and changed `IlastikPretrainedProbability` to clarify how to use in Windows and accept variable dataset names that are exported from ilastik.

[link to readthedocs](https://spacetx-starfish.readthedocs.io/en/mcai-segmentation-docs/creating_an_image_processing_pipeline/index.html#segmenting-cells)

### CODE REVIEW

File: docs/source/creating_an_image_processing_pipeline/index.rst [hunk 1]
```diff
@@ -142,6 +148,88 @@ Finding and Decoding Spots
 Segmenting Cells
 ----------------
 
+Unlike single-cell RNA sequencing, image-based transcriptomics methods do not physically separate
+cells before acquiring RNA information. Therefore in order to characterize cells, the RNA must be
```
Comment (ttung) [inline docs/source/creating_an_image_processing_pipeline/index.rst]:
```suggestion
cells before acquiring RNA information. Therefore, in order to characterize cells, the RNA must be
```

File: docs/source/creating_an_image_processing_pipeline/index.rst [hunk 2]
```diff
@@ -142,6 +148,88 @@ Finding and Decoding Spots
...
+assigned into single cells by partitioning the image volume. Accurate unsupervised cell-segmentation
+is an `open problem <https://www.kaggle.com/c/data-science-bowl-2018>`_ for all biomedical imaging
+disciplines ranging from digital pathology to neuroscience.
+
+The challenge of segmenting cells depends on the structural complexity of the sample and quality
+of images available. For example a sparse cell mono-layer with a strong cytosol stain would be
```
Comment (ttung) [inline docs/source/creating_an_image_processing_pipeline/index.rst]:
```suggestion
of images available. For example, a sparse cell mono-layer with a strong cytosol stain would be
```

File: docs/source/creating_an_image_processing_pipeline/index.rst [hunk 3]
```diff
@@ -142,6 +148,88 @@ Finding and Decoding Spots
...
+trivial to segment but a dense heterogeneous population of cells in 3D tissue with only a DAPI stain
+can be impossible to segment perfectly. On the experimental side, selecting good cell stains and
+acquiring images with low background will make segmenting a more tractable task.
+
+There are many approaches for segmenting cells from image-based transcriptomics assays. Below are
+a few methods that are implemented or integrated with starfish to output a
+:py:class:`.BinaryMaskCollection`, which represents a collection of labeled objects. If you do not
+know which segmentation method to use, a safe bet is to start with thresholding and watershed. On
+the other hand, if you can afford to manually define ROI masks there is no better way to
```
Comment (ttung) [inline docs/source/creating_an_image_processing_pipeline/index.rst]: is ROI common enough parlance that it's not necessary to define it?

File: docs/source/creating_an_image_processing_pipeline/index.rst [hunk 4]
```diff
@@ -142,6 +148,88 @@ Finding and Decoding Spots
...
+guarantee accurate segmentation.
+
+.. note::
+    While there is no "ground truth" for cell segmentation the closest approximation is manual
```
Comment (ttung) [inline docs/source/creating_an_image_processing_pipeline/index.rst]:
```suggestion
    While there is no "ground truth" for cell segmentation, the closest approximation is manual
```

File: docs/source/creating_an_image_processing_pipeline/index.rst [hunk 5]
```diff
@@ -142,6 +148,88 @@ Finding and Decoding Spots
...
+
+When overlapping cells are labeled as one connected component, they are typically segmented by
+using a `distance transformation followed by the watershed algorithm <https://www.mathworks
+.com/company/newsletters/articles/the-watershed-transform-strategies-for-image-segmentation
+.html>`_. Watershed is a classic image processing algorithm for separating objects in images and
+can be applied to all types of images. Pairing it with a distance transform is particularly
+useful for segmenting convex shapes like cells.
+
+A segmentation pipeline that consists of thresholding, connected component analysis, and watershed
+is the simplest and fastest to implement but its accuracy is highly dependent on image quality.
```
Comment (ttung) [inline docs/source/creating_an_image_processing_pipeline/index.rst]:
```suggestion
is simple and fast to implement but its accuracy is highly dependent on image quality.
```

File: docs/source/creating_an_image_processing_pipeline/index.rst [hunk 6]
```diff
@@ -142,6 +148,88 @@ Finding and Decoding Spots
...
+segmentation class that uses the :term:`primary images<Primary Images>` as the cell stain.
+
+* Tutorial: :ref:`Ways to segment by thresholding and watershed<tutorial_watershed_segmentation>`
+
+Manually Defining Cells
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The most accurate but time-consuming approach is to manually segment images using a tool such as
+`ROI manager <https://imagej.net/docs/guide/146-30.html#fig:The-ROI-Manager>`_ in FIJI (ImageJ). It
+is a straightforward process that starfish supports by allowing ROI set binaries to be imported as a
```
Comment (ttung) [inline docs/source/creating_an_image_processing_pipeline/index.rst]: "ROI set binaries" is confusing to me.  Do you mean ROI sets saved as binary files?

File: docs/source/creating_an_image_processing_pipeline/index.rst [hunk 7]
```diff
@@ -142,6 +148,88 @@ Finding and Decoding Spots
...
+methods that aim to replicate the accuracy of manual cell segmentation while reducing the labor
+required. Machine-learning algorithms for segmentation are continually improving but there is no
+perfect solution for all image types yet. These methods require training data (e.g. stained
+images with manually defined labels) to train a model to predict cell or nuclei locations in test
+data. There are `exceptions that don't require training on your specific data <http://www.cellpose
+.org/>`_ but generally training the model is something to consider when evaluating how much time
+each segmentation approach will require.
+
+Starfish currently has built-in functionality to support `ilastik <https://www.ilastik.org/>`_, a
+segmentation toolkit that leverages machine-learning. Ilastik has a Pixel Classification
```
Comment (ttung) [inline docs/source/creating_an_image_processing_pipeline/index.rst]: Is there a link to Ilastik's pixel classification workflows that might be pertinent here?

File: examples/ilastik_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+a segmentation toolkit that leverages machine-learning. Ilastik has a Pixel Classification
+workflow that performs semantic segmentation of the image, returning probability maps for each
+label (e.g. cells). To use ilastik in starfish first install ilastik locally and follow the
+Pixel Classification workflow using the GUI to train a classifier, which is saved in an ilastik
+project file.
+
+There are two options for using ilastik in starfish. The first is running the trained pixel
+classifier on images outside of starfish to generate a probability map that can then be loaded in
+starfish as an :py:class:`.ImageStack`. The second is to run the pixel classifier from within
+starfish by calling out to ilastik and passing it the trained classifier. The results will be the
```
Comment (ttung) [inline examples/ilastik_segmentation.py]:
```suggestion
starfish by calling out to ilastik and passing it to the trained classifier. The results will be the
```

File: examples/manual_segmentation.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+Loading ImageJ ROI Sets
+=======================
+
+.. note::
+    Starfish only supports importing 2D ROIs.
+
+In order to create a cell by gene expression matrix from image-based transcriptomics data, RNA
+spots must be assigned to cells by segmenting an image. The best quality cell segmentation
+annotations are manually drawn by experts. If you have ROI sets exported with `ROI manager
+<https://imagej.net/docs/guide/146-30.html#fig:The-ROI-Manager>`_ in ImageJ or FIJI they can be
```
Comment (ttung) [inline examples/manual_segmentation.py]:
```suggestion
<https://imagej.net/docs/guide/146-30.html#fig:The-ROI-Manager>`_ in ImageJ or FIJI, they can be
```

File: docs/source/creating_an_image_processing_pipeline/index.rst (hunk 3 above)
Comment (mattcai) [reply docs/source/creating_an_image_processing_pipeline/index.rst]: Good point, I will add an entry for ROI in the glossary and link to it.

File: docs/source/creating_an_image_processing_pipeline/index.rst (hunk 6 above)
Comment (mattcai) [reply docs/source/creating_an_image_processing_pipeline/index.rst]: Yes that's what I meant... I got the language from some ROI manager documentation I can't find now but I agree it's confusing. How about "that starfish supports by importing ROI sets stored in ZIP archives".

File: docs/source/creating_an_image_processing_pipeline/index.rst (hunk 7 above)
Comment (mattcai) [reply docs/source/creating_an_image_processing_pipeline/index.rst]: Yes added

File: examples/ilastik_segmentation.py (hunk 8 above)
Comment (mattcai) [reply examples/ilastik_segmentation.py]: My understanding is that the trained classifier is passed to ilastik (the "it" in my sentence). Tony's comment makes "it" the image stack, which is passed to the trained classifier. Which is correct?

`subprocess.check_call(
                [self.ilastik_executable,
                 '--headless',
                 '--project',
                 self.ilastik_project,
                 "--output_filename_format",
                 output_file,
                 dapi_file], env={})
`

Comment (ttung) [reply examples/ilastik_segmentation.py]: Sorry, you're right.  Original sentence is better.

File: docs/source/creating_an_image_processing_pipeline/index.rst (hunk 6 above)
Comment (ttung) [reply docs/source/creating_an_image_processing_pipeline/index.rst]: sounds good.

**Pull request merged**